source/promise-race-with-signal.ts

```diff
@@ -8,19 +8,48 @@ import {promiseFromSignal} from './promise-from-signal.js';
  * @param signal The signal to listen to. If you pass a controller, it will automatically extract its signal.
  * @param options.abortRejects If `false`, the promise will resolve instead of reject when the signal is aborted.
  */
-async function promiseRaceWithSignal<T, AbortRejects extends boolean = true>(
+async function promiseRaceWithSignal<T>(
 	promise: Promise<T>,
 	signal: AbortSignal | AbortController | undefined,
-	{abortRejects = true as AbortRejects}: {abortRejects?: AbortRejects} = {},
-): Promise<AbortRejects extends false ? T | unknown : T> {
+	options: {abortRejects?: false}
+): Promise<T | unknown>;
+
+async function promiseRaceWithSignal<T>(
+	promise: Promise<T>,
+	signal: AbortSignal | AbortController | undefined,
+	options?: {abortRejects: true}
+): Promise<T>;
+
+async function promiseRaceWithSignal<T>(
+	promise: Promise<T>,
+	signal: AbortSignal | AbortController | undefined,
+	{abortRejects = true}: {abortRejects?: boolean} = {},
+): Promise<T | unknown> {
 	if (!signal) {
 		return promise;
 	}
 
-	return Promise.race([
-		promise,
-		promiseFromSignal(signal, {rejects: abortRejects}),
-	]) as Promise<AbortRejects extends false ? T | unknown : T>;
+	const trueSignal = signal instanceof AbortController ? signal.signal : signal;
+	if (abortRejects) {
+		trueSignal.throwIfAborted();
+	} else if (trueSignal.aborted) {
+		return trueSignal.reason;
+	}
+
+	return new Promise((resolve, reject) => {
+		const handleAbort = () => {
+			if (abortRejects) {
+				reject(trueSignal.reason);
+			} else {
+				resolve(trueSignal.reason);
+			}
+		};
+
+		trueSignal.addEventListener('abort', handleAbort, {once: true});
+		promise.then(resolve, reject).finally(() => {
```

Comment:
Wouldn't this cause an uncaught error if the signal aborted first because it will call resolve/reject a 2nd time?

Reply:
Promises can only be fulfilled once, so successive resolve/reject calls are noops

<img width="350" alt="Screenshot" src="https://github.com/user-attachments/assets/69c5657f-e38e-466f-a73c-32c0b95f1597" />